org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/navigation/NavigationService.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2016, 2020 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ui.editor.navigation;
+
+import java.util.function.Consumer;
+
+import org.eclipse.jface.util.OpenStrategy;
+import org.eclipse.jface.viewers.ISelection;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.jface.viewers.OpenEvent;
+import org.eclipse.jface.viewers.StructuredViewer;
+import org.eclipse.ui.OpenAndLinkWithEditorHelper;
+import org.eclipse.xtext.ide.editor.navigation.INavigatable;
+import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.IReferenceDescription;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.ui.editor.IURIEditorOpener;
+import org.eclipse.xtext.xbase.lib.Procedures.Procedure1;
+
+import com.google.common.collect.Streams;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * @since 2.10
+ * 
+ * @author kosyakov - Initial contribution and API
+ */
+@Singleton
+public class NavigationService {
+
+	@Inject
+	private IURIEditorOpener uriEditorOpener;
+
+	public void open(final OpenEvent openEvent) {
+		open(openEvent, true);
+	}
+
+	@SuppressWarnings("unchecked")
+	public void open(OpenEvent openEvent, boolean select) {
+		final ISelection selection = openEvent.getSelection();
+		if (selection instanceof IStructuredSelection) {
+			Streams.stream(((IStructuredSelection) selection).iterator()) //
```

Comment:
A little stream-api-happy here ... a simple for-loop would be easier to read / debug. Not a request to rewrite this now, though.